core/src/latest_unprocessed_votes.rs

```diff
@@ -0,0 +1,685 @@
+    packet_indexes.iter().filter_map(move |packet_index| {
+        LatestValidatorVotePacket::new(packet_batch[*packet_index].clone(), vote_source).ok()
+    })
+}
+
+// TODO: replace this with rand::seq::index::sample_weighted once we can update rand to 0.8+
+// This requires updating dependencies of ed25519-dalek as rand_core is not compatible cross
+// version https://github.com/dalek-cryptography/ed25519-dalek/pull/214
+pub(crate) fn weighted_random_order_by_stake<'a>(
+    bank: &Arc<Bank>,
+    pubkeys: impl Iterator<Item = &'a Pubkey>,
+) -> impl Iterator<Item = Pubkey> {
+    // Efraimidis and Spirakis algo for weighted random sample without replacement
+    let staked_nodes = bank.staked_nodes();
+    let mut pubkey_with_weight: Vec<(f64, Pubkey)> = pubkeys
+        .filter_map(|&pubkey| {
+            let stake = staked_nodes.get(&pubkey).copied().unwrap_or(0);
+            if stake == 0 {
+                None // Ignore votes from unstaked validators
+            } else {
+                Some((thread_rng().gen::<f64>().powf(1.0 / (stake as f64)), pubkey))
+            }
+        })
+        .collect::<Vec<_>>();
+    pubkey_with_weight.sort_by(|(w1, _), (w2, _)| w1.partial_cmp(w2).unwrap());
+    pubkey_with_weight.into_iter().map(|(_, pubkey)| pubkey)
+}
+
+pub struct VoteBatchInsertionMetrics {
+    num_dropped_gossip: usize,
+    num_dropped_tpu: usize,
+}
+
+#[derive(Debug, Default)]
+pub struct LatestUnprocessedVotes {
+    latest_votes_per_pubkey: RwLock<HashMap<Pubkey, Arc<RwLock<LatestValidatorVotePacket>>>>,
+}
+
+impl LatestUnprocessedVotes {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Expensive because this involves iterating through and locking every unprocessed vote
+    pub fn len(&self) -> usize {
+        self.latest_votes_per_pubkey
+            .read()
+            .unwrap()
+            .values()
+            .filter(|lock| !lock.read().unwrap().is_vote_taken())
+            .count()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn insert_batch(
+        &self,
+        votes: impl Iterator<Item = LatestValidatorVotePacket>,
+    ) -> VoteBatchInsertionMetrics {
+        let mut num_dropped_gossip = 0;
+        let mut num_dropped_tpu = 0;
+
+        for vote in votes {
+            if let Some(vote) = self.update_latest_vote(vote) {
+                match vote.vote_source {
+                    VoteSource::Gossip => num_dropped_gossip += 1,
+                    VoteSource::Tpu => num_dropped_tpu += 1,
+                }
+            }
+        }
+
+        VoteBatchInsertionMetrics {
+            num_dropped_gossip,
+            num_dropped_tpu,
+        }
+    }
+
+    fn get_entry(&self, pubkey: Pubkey) -> Option<Arc<RwLock<LatestValidatorVotePacket>>> {
+        self.latest_votes_per_pubkey
+            .read()
+            .unwrap()
+            .get(&pubkey)
+            .cloned()
```

Comment:
Just double checking, there's no danger of this `Arc<RwLock>` for this *same* pubkey being replaced with a new one right?

i.e. something like

Read `A`
`A` gets removed from the hashmap
`A'` gets re-inserted in to the hashmap
Some other thread gets `A'`

Reply:
I don't think that's possible, there's no operations to remove an entry, only clear the packet it contains. Also we only add new pubkeys if they don't currently exist in the map